server/services/invidious.ts

```diff
@@ -0,0 +1,426 @@
+			});
+			data = resLocal.data;
+		} catch {
+			// Some instances gate `local=1` behind anti-DDoS/rate-limit checks; fallback to non-proxied metadata. – Invidious infra quirk
+			const res = await this.api.get<InvidiousApiVideo>(baseUrl, {
+				headers: { Accept: "application/json" },
+			});
+			data = res.data;
+		}
+
+		if (!data?.title) {
+			throw new Error(`Invidious API returned empty/invalid response for ${host}:${id}`);
+		}
+
+		// Prefer DASH/HLS (higher quality) and fall back to progressive "direct"
+		const video = await this.parseAsDirect(data, host, id);
+
+		// Pre-fill cache to avoid a later HEAD/probe by the DirectVideoAdapter that might be blocked by CORS
+		// or upstream shields on public instances. – Browser+CORS / Invidious infra
+		try {
+			await storage.updateVideoInfo(video);
+		} catch (e: unknown) {
+			const msg = e instanceof Error ? e.message : String(e);
+			log.warn(`Failed to prefill cache for ${video.service}:${video.id}: ${msg}`);
+		}
+		return video;
+	}
+
+	/** Proxied Progressive (MP4/WebM) over Invidious – avoids YouTube signature churn and ensures CORS-compatible URLs. – YouTube+Invidious behavior */
+	private proxiedProgressive(host: string, id: string, itag?: number) {
+		const u = new URL(`https://${host}/latest_version`);
+		u.searchParams.set("id", id);
+		if (itag) {
+			u.searchParams.set("itag", String(itag));
+		}
+		u.searchParams.set("local", "1");
+		// Helps some instances route the request correctly. – Invidious-specific
+		u.searchParams.set("source", "youtube");
+		return u.toString();
+	}
+
+	private pickBestThumbnail(
+		thumbnails?: { url: string; width?: number; height?: number }[]
+	): string | undefined {
+		if (!thumbnails || thumbnails.length === 0) {
+			return undefined;
+		}
+		// YouTube commonly exposes multiple thumbnail sizes; prefer the largest. – YouTube behavior
+		const sorted = [...thumbnails].sort((a, b) => {
+			const aPixels = (a.width || 0) * (a.height || 0);
+			const bPixels = (b.width || 0) * (b.height || 0);
+			return bPixels - aPixels;
+		});
+		return sorted[0]?.url;
+	}
+
+	private async parseAsDirect(inv: InvidiousApiVideo, host: string, id: string): Promise<Video> {
+		const rawDesc = (inv.description ?? inv.shortDescription ?? "").toString().trim();
+		const safeDesc = rawDesc.length ? rawDesc : inv.title;
+
+		const base = {
+			title: inv.title,
+			description: safeDesc,
+			length:
+				typeof inv.lengthSeconds === "string"
+					? parseInt(inv.lengthSeconds, 10)
+					: inv.lengthSeconds,
+			thumbnail: this.pickBestThumbnail(inv.videoThumbnails),
+		};
+
+		// 1) Probe DASH & HLS and pick the manifest whose top variant has the higher bitrate.
+		try {
+			// Both may exist or be omitted depending on upstream availability and instance config. – YouTube/Invidious variability
+			const [dashProbe, hlsProbe] = await Promise.all([
+				this.probeManifest(host, id, "dash"),
+				this.probeManifest(host, id, "hls"),
+			]);
```

Comment:
For this case, if one of the probe fails, then the entire operation fails. This is because `Promise.all` fails if any of the promises it's handling fails (see: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all)

Your comment suggests that either of these could fail independently, hence my concern. We only want to outright fail if we can do none of the probes. I think you might want `Promise.any` instead